File: backend/app/models/posgrado_indicator.py
from sqlalchemy import Column, Integer, String, Numeric, UniqueConstraint

from app.database.database import Base


class PosgradoIndicator(Base):
    __tablename__ = "posgrado_indicators"

    id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    periodo = Column(String(10), nullable=False, index=True)
    tipo_programa = Column(String(20), nullable=False)

    matriculados = Column(Integer, nullable=False, default=0)
    graduados = Column(Integer, nullable=False, default=0)
    desertores = Column(Integer, nullable=False, default=0)
    promedio_acumulado = Column(Numeric(4, 2), nullable=False, default=0)

    hombres = Column(Integer, nullable=False, default=0)
    mujeres = Column(Integer, nullable=False, default=0)

    empleados = Column(Integer, nullable=False, default=0)
    desempleados = Column(Integer, nullable=False, default=0)

    financiacion_propia = Column(Integer, nullable=False, default=0)
    financiacion_beca = Column(Integer, nullable=False, default=0)
    financiacion_credito = Column(Integer, nullable=False, default=0)

    ponencias = Column(Integer, nullable=False, default=0)
    publicaciones = Column(Integer, nullable=False, default=0)

    proyectos_desarrollo = Column(Integer, nullable=True)
    proyectos_finalizados = Column(Integer, nullable=True)

    __table_args__ = (
        UniqueConstraint("periodo", "tipo_programa", name="uq_posgrado_periodo_tipo"),
    )

File: frontend/src/models/PosgradoIndicator.ts
export interface PosgradoIndicator {
  id: number;
  periodo: string;
  tipo_programa: string;
  matriculados: number;
  graduados: number;
  desertores: number;
  promedio_acumulado: number;
  hombres: number;
  mujeres: number;
  empleados: number;
  desempleados: number;
  financiacion_propia: number;
  financiacion_beca: number;
  financiacion_credito: number;
  ponencias: number;
  publicaciones: number;
  proyectos_desarrollo: number | null;
  proyectos_finalizados: number | null;
}

export interface PosgradoIndicatorStats {
  matriculados: number;
  graduados: number;
  desertores: number;
  promedio_acumulado: number;
  empleados: number;
  desempleados: number;
  ponencias: number;
  publicaciones: number;
  proyectos_desarrollo: number | null;
  proyectos_finalizados: number | null;
}

export interface PosgradoGenderStats {
  hombres: number;
  mujeres: number;
}

export interface PosgradoTrendDataPoint {
  periodo: string;
  matriculados: number;
  graduados: number;
  desertores: number;
}

export interface PosgradoFinancingStats {
  propia: number;
  beca: number;
  credito: number;
}

export interface PosgradoComputedStats {
  tasa_sobrepermanencia: number;
  tasa_deserciones: number;
  tasa_retirados_bra: number;
  tasa_graduados_10: number;
  tasa_graduados_mas_10: number;
}

export type PosgradoProgramType = 'MAESTRIA' | 'ESPECIALIZACION';
